src/software/robot_diagnostics/widgets/drive.cpp

```diff
@@ -0,0 +1,113 @@
+    };
+    setupSliderLineEdit(widget->lineEdit_direct_per_wheel_fl,
+                        widget->slider_direct_per_wheel_fl,
+                        direct_per_wheel_power_fl_changed, -100.0, 100.0, 1.0);
+    setupSliderLineEdit(widget->lineEdit_direct_per_wheel_fr,
+                        widget->slider_direct_per_wheel_fr,
+                        direct_per_wheel_power_fr_changed, -100.0, 100.0, 1.0);
+    setupSliderLineEdit(widget->lineEdit_direct_per_wheel_bl,
+                        widget->slider_direct_per_wheel_bl,
+                        direct_per_wheel_power_bl_changed, -100.0, 100.0, 1.0);
+    setupSliderLineEdit(widget->lineEdit_direct_per_wheel_br,
+                        widget->slider_direct_per_wheel_br,
+                        direct_per_wheel_power_br_changed, -100.0, 100.0, 1.0);
+
+    auto direct_per_wheel_stop_pushed = [widget,
+                                         direct_per_wheel_power_changed_callback]() {
+        // Reset sliders
+        widget->lineEdit_direct_per_wheel_fl->setText(QString::number(0));
+        widget->slider_direct_per_wheel_fl->setValue(0);
+        widget->lineEdit_direct_per_wheel_fr->setText(QString::number(0));
+        widget->slider_direct_per_wheel_fr->setValue(0);
+        widget->lineEdit_direct_per_wheel_bl->setText(QString::number(0));
+        widget->slider_direct_per_wheel_bl->setValue(0);
+        widget->lineEdit_direct_per_wheel_br->setText(QString::number(0));
+        widget->slider_direct_per_wheel_br->setValue(0);
+
+        direct_per_wheel_power_changed_callback(0, DirectPerWheelMode::FRONT_LEFT);
+        direct_per_wheel_power_changed_callback(0, DirectPerWheelMode::FRONT_RIGHT);
+        direct_per_wheel_power_changed_callback(0, DirectPerWheelMode::BACK_LEFT);
+        direct_per_wheel_power_changed_callback(0, DirectPerWheelMode::BACK_RIGHT);
+    };
+
+    QWidget::connect(widget->pushButton_direct_per_wheel_stop, &QPushButton::clicked,
+                     direct_per_wheel_stop_pushed);
+}
+
+void setupDirectVelocity(Ui::AutoGeneratedMainWidget* widget,
+                         std::function<void(double, DirectVelocityMode)>
+                             direct_velocity_power_changed_callback)
+{
+    auto direct_velocity_x_vel_changed =
+        [direct_velocity_power_changed_callback](double value) {
+            direct_velocity_power_changed_callback(value, DirectVelocityMode::X_VEL);
+        };
+    auto direct_velocity_y_vel_changed =
+        [direct_velocity_power_changed_callback](double value) {
+            direct_velocity_power_changed_callback(value, DirectVelocityMode::Y_VEL);
+        };
+    auto direct_velocity_angular_vel_changed = [direct_velocity_power_changed_callback](
+                                                   double value) {
+        direct_velocity_power_changed_callback(value, DirectVelocityMode::ANGULAR_VEL);
+    };
+
+    setupSliderLineEdit(widget->lineEdit_direct_velocity_x,
```

Comment:
can we add units to all of these, its hard to tell why some are 100 and some are a 1000

here as well as on the corresponding UI section

the old MRF test didn't have them but it would be nice to do it now

Reply:
added a comment